tektoncd/pipeline
# WIP [TEP-0139] Trusted Artifacts - POC

# Changes

This PR includes three commits and a partial implementation of the trusted artifacts POC.
The types are defined at this stage, and artifact parameters are ~functional.
Artifact results are not yet functional, so I added an example pipeline run which can be used
to test things out:
- The producer includes the step that will later be injected
- The consumer uses syntactic sugar to get the download and verification step injected

For now the injected steps are implemented as bash scripts, which are quick to develop
and have the advantage of letting users see what's going on, hopefully helping with
debuggability, which was one of the issues with `PipelineResources`.

In future we may replace those shell scripts with a simple go app or python scripts, this is
something we can discuss in the TEP design phase.

#### [TEP-0139] Introduce type artifact
   
    Introduce params and results of type "artifact". This type is
    identical to the type "object" but it provides a default
    object schema (path, hash, type).

    The artifact type only applies to the spec. ParamValues for
    artifact types are of type object.

    This is a first step in the implementation of TEP-0139.
    Future PRs will use artifact type params and results to
    automatically inject upload/download/hash/verify steps.

####  [TEP-0139] Add artifact workspaces

    Add "artifact" to pipeline and task workspaces. Add validation
    that only one workspace per task and one workspace per pipeline
    is marked as "artifact".

####  [TEP-0139] Add artifact download and verify step

    Add the artifact internal package and the function to generate
    an artifact download and verify step. This is used by the
    taskrun controller to inject a step which downloads and
    verify any param of type artifact before the actual steps in
    the task may used them.

    There are a number of TODOs in the code to be addressed
    and a sample pipeline in trusted_artifacts_halfway.yaml which
    can be used to try things out.

    Signed-off-by: Andrea Frittoli <andrea.frittoli@uk.ibm.com>

Example run of the pipeline:

```
? Select pipelinerun: trusted-artifacts-half-s7c8p started 2 seconds ago
[producer : produce-file] 1691676416

[producer : upload-hash-file] + ARTIFACT_ROOT=/tekton/artifacts
[producer : upload-hash-file] + A_FILE_PATH=afile.txt
[producer : upload-hash-file] ++ awk '{ print $1 }'
[producer : upload-hash-file] ++ md5sum /tekton/artifacts/afile.txt
[producer : upload-hash-file] + A_FILE_HASH=e2c9dea8707a572b81e30db1ee759399
[producer : upload-hash-file] + TARGET_PATH=/workspace/artifactStorage/.tekton/artifacts
[producer : upload-hash-file] + mkdir -p /workspace/artifactStorage/.tekton/artifacts
[producer : upload-hash-file] + cp /tekton/artifacts/afile.txt /workspace/artifactStorage/.tekton/artifacts/e2c9dea8707a572b81e30db1ee759399
[producer : upload-hash-file] + cat
[producer : upload-hash-file] + tee /tekton/results/aFileArtifact
[producer : upload-hash-file] {
[producer : upload-hash-file]   "path": "afile.txt",
[producer : upload-hash-file]   "hash": "e2c9dea8707a572b81e30db1ee759399",
[producer : upload-hash-file]   "type": "file"
[producer : upload-hash-file] }

[producer : produce-folder] 1691676417
[producer : produce-folder] 1691676417
[producer : produce-folder] 1691676417

[producer : upload-hash-folder] + ARTIFACT_ROOT=/tekton/artifacts
[producer : upload-hash-folder] + A_FOLDER_PATH=afolder
[producer : upload-hash-folder] + tar zcf /tekton/artifacts/afolder.tgz /tekton/artifacts/afolder
[producer : upload-hash-folder] tar: removing leading '/' from member names
[producer : upload-hash-folder] ++ md5sum /tekton/artifacts/afolder.tgz
[producer : upload-hash-folder] ++ awk '{ print $1 }'
[producer : upload-hash-folder] + A_FOLDER_HASH=356c5e96e92607798b15a63414a6a3c3
[producer : upload-hash-folder] + TARGET_PATH=/workspace/artifactStorage/.tekton/artifacts
[producer : upload-hash-folder] + mkdir -p /workspace/artifactStorage/.tekton/artifacts
[producer : upload-hash-folder] + cp /tekton/artifacts/afolder.tgz /workspace/artifactStorage/.tekton/artifacts/356c5e96e92607798b15a63414a6a3c3.tgz
[producer : upload-hash-folder] + cat
[producer : upload-hash-folder] + tee /tekton/results/aFolderArtifact
[producer : upload-hash-folder] {
[producer : upload-hash-folder]   "path": "afolder",
[producer : upload-hash-folder]   "hash": "356c5e96e92607798b15a63414a6a3c3",
[producer : upload-hash-folder]   "type": "folder"
[producer : upload-hash-folder] }

[consumer : tekton-artifact-download] /tekton/artifacts/afile.txt: OK
[consumer : tekton-artifact-download] /tekton/artifacts/afolder.tgz: OK

[consumer : consume-content] /tekton/artifacts/afile.txt
[consumer : consume-content] /tekton/artifacts/afolder/tekton/artifacts/afolder/a.txt
[consumer : consume-content] /tekton/artifacts/afolder/tekton/artifacts/afolder/b.txt
[consumer : consume-content] /tekton/artifacts/afolder/tekton/artifacts/afolder/c.txt
```

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
WIP [TEP-0139], not fully functional yet.
```

/kind feature
